bpo-31336: Speed up type creation, which is highly dominated by slow dict lookups.

scoder opened the pull request:
This gives me around 20% better performance for ``class Test: pass``. Admittedly, that's a micro-benchmark, but the change is obvious enough to not merit staying slow.

<!-- issue-number: bpo-31336 -->
https://bugs.python.org/issue31336
<!-- /issue-number -->

Commit:
Speed up type creation, which is highly dominated by slow dict lookups.

serhiy-storchaka commented on Objects/typeobject.c:
```diff
@@ -2983,6 +2984,21 @@ _PyType_Lookup(PyTypeObject *type, PyObject *name)
+
     res = NULL;
     /* keep a strong reference to mro because type->tp_mro can be replaced
        during PyDict_GetItem(dict, name)  */
```
Update the comment.

serhiy-storchaka commented on Objects/typeobject.c:
```diff
@@ -2994,7 +3010,7 @@ _PyType_Lookup(PyTypeObject *type, PyObject *name)
         dict = ((PyTypeObject *)base)->tp_dict;
         assert(dict && PyDict_Check(dict));
-        res = PyDict_GetItem(dict, name);
+        res = _PyDict_GetItem_KnownHash(dict, name, hash);
```
`PyDict_GetItem()` always clears errors. Call `PyErr_Clear()` if `_PyDict_GetItem_KnownHash()` returns NULL.

scoder replied:
Already changed.

vstinner replied:
I don't see well the difference between PyDict_GetItem() and _PyDict_GetItem_KnownHash() in term of performance. Maybe the difference is that PyDict_GetItem() calls PyErr_Fetch/PyErr_Restore. In that case, PyDict_GetItemWithError() would have the same speed, no?

serhiy-storchaka replied:
Using `PyDict_GetItemWithError()` adds only a half of the speed up.

The difference is:
1. PyDict_GetItem() calls PyThreadState_GET/PyErr_Fetch/PyErr_Restore.
2. PyDict_GetItem() checks for str and reads a hash every time. I don't understand well why this affects performance.

scoder replied:
KnownHash is extremely short in comparison and probably gets inlined and streamlined with LTO. Substantially less branching.

Commit:
Update comment after changing the function call that it refers to.

Commit:
Ignore any errors (however unlikely) that may happen during the mro chained name lookups in _PyType_Lookup().

scoder commented:
Nice catches, but certainly, I didn't expect the Spanish Inquisition!

vstinner reviewed (changes requested):
I would prefer to use the _Py_IDENTIFIER API rather than using _PyDict_GetItem_KnownHash().

Commit:
Extract non-caching code from _PyType_Lookup() to use it directly from update_one_slot().

Commit:
Avoid some useless overhead for the no-basetype case. If "object" ever ceases to be a valid base type, there'll probably be larger code sections to change than this one.

scoder commented:
I agree that using the `_Py_IDENTIFIER` API would be nice, but changing the whole setup is more work than I would currently like to invest into this. This can still be done later. My changes do not make it any harder.

Instead, I followed Naokis idea of spliting `_PyType_Lookup()` into two functions, one to do the lookup and one that handles the method caching and calls the other on misses.

pitrou commented on Objects/typeobject.c:
```diff
@@ -6958,8 +6990,14 @@ update_one_slot(PyTypeObject *type, slotdef *p)
     }
     do {
-        descr = _PyType_Lookup(type, p->name_strobj);
+        descr = _PyType_LookupUncached(type, p->name_strobj, &error);
```
Would be nice to add a comment explaining why the uncached lookup.

scoder replied:
Sure enough. Added.

Commit:
Add comment.

Commit:
Avoid uselessly searching empty bases for a metaclass. This is quite common in Python 3 now.
Also make the step from "return NULL" error handling to "goto error" reference cleanup explicit.

serhiy-storchaka commented on Objects/typeobject.c:
```diff
@@ -2360,35 +2360,39 @@ type_new(PyTypeObject *metatype, PyObject *args, PyObject *kwds)
     }
-    else
-        Py_INCREF(bases);
+    else {
```
Is there any effect of this change? I tested `class C: pass` and `class C(object): pass` and didn't see any difference.

scoder replied:
I consider this part a cleanup that makes it clearer what operations are needed when, and how error cases are dealt with.
I didn't measure any speed difference either.

serhiy-storchaka commented on Objects/typeobject.c:
```diff
@@ -2994,11 +2985,53 @@ _PyType_Lookup(PyTypeObject *type, PyObject *name)
         if (res != NULL)
             break;
+        if (PyErr_Occurred())
+            return NULL;
```
Leaks `mro`.

scoder replied:
Fixed.

serhiy-storchaka commented on Objects/typeobject.c:
```diff
@@ -2983,9 +2965,18 @@ _PyType_Lookup(PyTypeObject *type, PyObject *name)
+    if (!PyUnicode_CheckExact(name) ||
+        (hash = ((PyASCIIObject *) name)->hash) == -1)
+    {
+        hash = PyObject_Hash(name);
```
`PyObject_Hash()` can call user code. `mro` is a borrowed reference here, it should be increfed before calling `PyObject_Hash()`.

Or calculate the hash before getting `mro`.

scoder replied:
Good catch. I'll move the hash() call up.

serhiy-storchaka commented on Objects/typeobject.c:
```diff
@@ -2938,43 +2942,21 @@ PyType_GetSlot(PyTypeObject *type, int slot)
+/* Internal API to look for a name through the MRO, bypassing the method cache.
+   This returns a borrowed reference, and might set an exception! */
+static PyObject *
+_PyType_LookupUncached(PyTypeObject *type, PyObject *name, int *error)
```
Local functions usually use different name convention. All lower characters and no `Py` prefix.

scoder replied:
I was trying to hint at the name of the existing function. A different name would be less clear, I think.

serhiy-storchaka replied:
IMHO this rather adds a confusion.

scoder replied:
I renamed the function to `find_name_in_mro()`.

serhiy-storchaka commented on Objects/typeobject.c:
```diff
@@ -2938,43 +2942,21 @@ PyType_GetSlot(PyTypeObject *type, int slot)
-    }
-
+    Py_hash_t hash;
+    *error = 1;
```
Alternatively you could return a special value for signalling error. E.g. (PyObject *)(-1). Test what is faster.

scoder replied:
I can set `error` to `-1` on exceptions. That will distinguish all three cases: ok, error with exception, error without exception. Only `-1` will need a call to `PyErr_Clear()` then.

scoder replied:
I dislike things like `(PyObject *)(-1)` because I wouldn't want to assume that they are not valid pointer values on any system whatsoever.

Commit:
Avoid unsafe handling of borrowed "mro" reference during hash() call.
Distinguish between "error" and "error with exception" cases in _PyType_LookupUncached().
Fix a reference leak of "mro" on lookup errors.
Resolves issues found by Serhiy Storchaka.

scoder commented:
@serhiy-storchaka: thank you for your excellent feedback.
I applied the changes and couldn't see a performance degradation from them.

serhiy-storchaka commented on Objects/typeobject.c:
```diff
@@ -2965,27 +2967,20 @@ _PyType_Lookup(PyTypeObject *type, PyObject *name)
-               in a context that propagates the exception out.
-            */
-            PyErr_Clear();
+                PyType_Ready(type) < 0) {
```
Either move `PyType_Ready()` to the previous line (if the resulting line will be not too long), or indent it as in the current code and move `{` to a new line for readability (new PEP 7 rule).

scoder replied:
I cleaned it up a little.

serhiy-storchaka commented on Objects/typeobject.c:
```diff
@@ -2994,11 +2989,57 @@ _PyType_Lookup(PyTypeObject *type, PyObject *name)
+           in a context that propagates the exception out.
+        */
+        if (error == -1)
+            PyErr_Clear();
```
`PyErr_Clear()` always is called after `_PyType_LookupUncached()` if `error == -1`. Why not call it inside `_PyType_LookupUncached()`?

scoder replied:
The question is which interface is better: should the function itself always swallow any exceptions, or should the callers decide how to deal with them? The comments in the original `_PyType_Lookup()` implementation suggest that others have already been as unhappy as me about the interface of *that* function, so why keep making people unhappy?

serhiy-storchaka commented on Objects/typeobject.c:
```diff
@@ -2994,11 +2989,57 @@ _PyType_Lookup(PyTypeObject *type, PyObject *name)
+           the same type will call it again -- hopefully
+           in a context that propagates the exception out.
+        */
+        if (error == -1)
```
Add braces.

scoder replied:
Done.

Commit:
Clean up code and formatting a little.

Commit:
Add braces for code style reasons.

Commit:
Give internal helper function a local name that does not resemble (existing) API names.
Suggested by Serhiy Storchaka.

scoder commented:
PR updated.

vstinner commented on Objects/typeobject.c:
```diff
@@ -2994,11 +2990,58 @@ _PyType_Lookup(PyTypeObject *type, PyObject *name)
+        if (PyErr_Occurred()) {
+            *error = -1;
+            goto done;
+        }
```
I dislike error handling here. I suggest to call PyObject_GetItem() if PyDict_CheckExact() is false, and to really report the error to the caller (not ignore it). Would it possible?

I don't think that it's correct to call  _PyDict_GetItem_KnownHash() if tp_dict is a dict subtype.

scoder replied:
Also a good catch. That complicates things a bit, though. If we have a PyUnicode subtype for the name and a non-dict mapping, we could end up calling hash() once more than needed. I've moved the hash calculation down to get a lazy calculation for the exact-dict case. This leads to a small but measurable performance degradation of about 1-2% less speedup. It's still pretty good in comparison, I think.

vstinner reviewed (dismissed):
I don't have a strong opinion on Py_IDENTIFIER(). 10% speedup is nice to have. If it's not faster with Py_IDENTIFIER(), it's ok to use _PyDict_GetItem_KnownHash().

The overchange change LGTM, but IMHO the error handling and handling of types other than dict (including dict subclasses) is a blocker issue. See my comment.

vstinner commented:
FYI ceval.c was modified to handle namespaces which are not exactly the dict type, even if I'm not sure that this code is well tested :-)

For types, it see many use cases for dict subclasses or even other mapping types which don't inherit from dict, like proxies.

Commit:
Allow non-dict types for the class dict when looking up names and call PyObject_GetItem() for them.

Commit:
Lazily calculate name hash in find_name_in_mro() to avoid potential re-calculation if the mapping is not exactly a dict.

scoder commented:
PR updated to correctly support non-dict namespaces.

serhiy-storchaka commented on Objects/typeobject.c:
```diff
@@ -2938,67 +2942,113 @@ PyType_GetSlot(PyTypeObject *type, int slot)
-        res = PyDict_GetItem(dict, name);
+        assert(dict);
+        /* Optimise for the extremely common case: dict for lookup, unicode name. */
+        if (PyDict_CheckExact(dict)) {
```
This is overgeneralization. Can `tp_dict` be not exact dict at all? I don't think this is possible. In many places concrete dict API is used with `tp_dict`. If you want to allow `tp_dict` be not exact dict, please open a separate issue for this.

serhiy-storchaka reviewed (changes requested):
Please revert recent changes.

Commit:
Revert "Lazily calculate name hash in find_name_in_mro() to avoid potential re-calculation if the mapping is not exactly a dict."

This reverts commit 1d52082bb544de02901f922242942253497cba58.

Commit:
Revert "Allow non-dict types for the class dict when looking up names and call PyObject_GetItem() for them."

This reverts commit 09e716a2e8eda777400da779c5a4f1b5e90375f6.

scoder commented:
Reverted. Anything else missing from this PR?

serhiy-storchaka commented:
The code LGTM. Add a news entry and find why AppVeyor build failed.

Commit:
add news entry for faster class creation

scoder commented:
News entry added.
Appveyor says this:
```
======================================================================
ERROR: test_subclass (test.test_xml_etree.TreeBuilderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\cpython\lib\test\test_xml_etree.py", line 2441, in test_subclass
    self._check_sample1_element(e)
  File "C:\projects\cpython\lib\test\test_xml_etree.py", line 2391, in _check_sample1_element
    self.assertEqual(e.tag, 'html')
AttributeError: 'NoneType' object has no attribute 'tag'
======================================================================
FAIL: test_dummy_builder (test.test_xml_etree.TreeBuilderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\cpython\lib\test\test_xml_etree.py", line 2413, in test_dummy_builder
    self.assertEqual(parser.close(), 42)
AssertionError: None != 42
----------------------------------------------------------------------
Ran 145 tests in 0.234s
FAILED (failures=1, errors=1)
```
I also get these locally, but not for the base revision in master. I'll investigate.

scoder commented:
The `XMLParser.__init__()` code in `_elementtree.c` does this:
```
    self->handle_start = PyObject_GetAttrString(target, "start");
    self->handle_data = PyObject_GetAttrString(target, "data");
    self->handle_end = PyObject_GetAttrString(target, "end");
    self->handle_comment = PyObject_GetAttrString(target, "comment");
    self->handle_pi = PyObject_GetAttrString(target, "pi");
    self->handle_close = PyObject_GetAttrString(target, "close");
    self->handle_doctype = PyObject_GetAttrString(target, "doctype");
    PyErr_Clear();
```
In the failing test, it should find `close` but not `pi`. Thus, it looks up `close` with a live `AttributeError` set from the `pi`lookup. Since `_PyDict_GetItem_KnownHash()` may or may not set an exception, we have to check for a live exception after calling it, and that finds the old exception of the last attribute lookup and decides that its own lookup failed.

The correct place to fix this is obviously `_elementtree.c` (please do), but if one module has this bug, it's unlikely to be the only one. What do you think? To me, it would feel wrong to allow this misbehaviour inside of CPython by explicitly handling it somehow.

scoder commented:
That being said, this can obviously be made to work again by ignoring all exceptions also in the new helper function, just as `_PyType_Lookup()` previously did. Sounds wrong to do that, but it keeps things working that worked before.

Commit:
Change nice interface of "find_name_in_mro()" to evil interface eating lookup exceptions, just like "_PyType_Lookup()" did previously.

scoder commented:
I pushed a fix that eats lookup exceptions, exactly like `_PyType_Lookup()` previously did.

serhiy-storchaka commented on Misc/NEWS.d/next/Core and Builtins/2017-09-13-12-04-23.bpo-31336.gi2ahY.rst:
```diff
@@ -0,0 +1,2 @@
+Speed up class creation by reducing overhead in the necessary special method
```
Add speed up estimation: 10-20%

serhiy-storchaka commented:
What was the cause of this failure? Could this failure be fixed by fixing ElementTree?

scoder commented:
See my comment above. Yes, ET should definitely be fixed. I pushed #3545.

Commit:
Mention amount of speedup in News entry.

scoder commented:
Feel free to revert 4efde8ea72195e987a7946985b22725b1d8f4e3a if you think that fixing cElementTree is enough here.

Commit:
Revert "Change nice interface of "find_name_in_mro()" to evil interface eating lookup exceptions, just like "_PyType_Lookup()" did previously."

This reverts commit 4efde8ea72195e987a7946985b22725b1d8f4e3a.

Commit:
Guard against external live exceptions when calling find_name_in_mro() from functions that swallow live exceptions.

scoder commented:
I've reverted that change and added `assert()`s instead. Fixing the behaviour of `_PyType_Lookup()` is a separate issue (I've created a ticket in the tracker).

scoder commented:
Note that both AppVeyor and Travis will be happy again once #3545 is merged.

vstinner commented on Objects/typeobject.c:
```diff
@@ -6957,9 +7004,18 @@ update_one_slot(PyTypeObject *type, slotdef *p)
+            if (error == -1) {
+                /* It is unlikely by not impossible that there has been an exception
+                   during lookup. Since this function originally expected no errors,
+                   we ignore them here in order to keep up the interface. */
```
I don't think that "Since this function originally expected no errors" matters here. Why not reporting the exception to the caller and handle it?

scoder replied:
I'm working on a patch that cleans up much of the mess around `_PyType_Lookup()`. See https://bugs.python.org/issue31465

vstinner commented:
"KnownHash is extremely short in comparison and probably gets inlined and streamlined with LTO. Substantially less branching." oh nice, good to know!

scoder commented on Objects/typeobject.c:
```diff
@@ -2938,54 +2942,46 @@ PyType_GetSlot(PyTypeObject *type, int slot)
-        mro = type->tp_mro;
         if (mro == NULL) {
+            *error = 1;
             return NULL;
```
Since this is the only non-exception raising error case left, I actually wonder if this is really an error case. If there is no MRO in a ready-ied type, isn't that just fine from the point of view of a lookup?

scoder commented:
From my POV, this is ready for merging.
All further changes regarding the general exception handling in `_PyType_Lookup()` are tracked by bpo-31465 and #3616.

vstinner commented on Objects/typeobject.c:
```diff
@@ -2360,35 +2360,39 @@ type_new(PyTypeObject *metatype, PyObject *args, PyObject *kwds)
+        bases = PyTuple_Pack(1, base);
         if (bases == NULL)
-            goto error;
+            return NULL;
```
PEP 7 nitpick: when you modify code, it's better to add { ... } to if blocks.

vstinner reviewed (commented):
LGTM. I just proposed a minor coding style change.

The change enhances error handling, nice.

I don't understand everything, but I trust Python test suite to make sure that the change doesn't break anything :-)
